test/e2e/magic_catalog.go

```diff
@@ -0,0 +1,280 @@
+		fileBasedCatalog: provider,
+		kubeClient:       kubeClient,
+		namespace:        namespace,
+		name:             catalogName,
+		configMapName:    catalogName + "-configmap",
+		serviceName:      catalogName + "-svc",
+		podName:          catalogName + "-pod",
+	}
+}
+
+func (c *magicCatalog) DeployCatalog(ctx context.Context) error {
+
+	catalogSource := c.makeCatalogSource()
+	resourcesInOrderOfDeployment := []k8scontrollerclient.Object{
+		c.makeConfigMap(),
+		c.makeCatalogSourcePod(),
+		c.makeCatalogService(),
+		catalogSource,
+	}
+
+	for _, res := range resourcesInOrderOfDeployment {
+		err := c.kubeClient.Create(ctx, res)
+		if err != nil {
+			return c.cleanUpAfter(ctx, err)
+		}
+	}
+
+	// wait for catalog source to become ready
+	err := waitFor(func() (bool, error) {
+		err := c.kubeClient.Get(ctx, k8scontrollerclient.ObjectKey{
+			Name:      catalogSource.GetName(),
+			Namespace: catalogSource.GetNamespace(),
+		}, catalogSource)
+
+		if err != nil || catalogSource.Status.GRPCConnectionState == nil {
+			return false, err
+		}
+
+		state := catalogSource.Status.GRPCConnectionState.LastObservedState
+
+		if state != catalogReadyState {
+			return false, nil
+		} else {
+			return true, nil
+		}
+	})
+
+	if err != nil {
+		return c.cleanUpAfter(ctx, err)
+	}
+
+	return nil
+}
+
+func (c *magicCatalog) UndeployCatalog(ctx context.Context) []error {
+	var errs []error = nil
+
+	resourcesInOrderOfDeletion := []k8scontrollerclient.Object{
+		c.makeCatalogSource(),
+		c.makeCatalogService(),
+		c.makeCatalogSourcePod(),
+		c.makeConfigMap(),
+	}
+
+	// try to delete all resourcesInOrderOfDeletion even if errors are
+	// encountered through deletion.
+	for _, res := range resourcesInOrderOfDeletion {
+		err := c.kubeClient.Delete(ctx, res)
+
+		// ignore not found errors
+		if err != nil && !k8serror.IsNotFound(err) {
+			if errs == nil {
+				errs = make([]error, 0)
+			}
+			errs = append(errs, err)
+		}
+	}
+
+	return errs
+}
+
+func (c *magicCatalog) cleanUpAfter(ctx context.Context, err error) error {
+	cleanupErr := c.UndeployCatalog(ctx)
+	if cleanupErr != nil {
+		return fmt.Errorf("the following cleanup errors occurred: '%s' after an error deploying the configmap: '%s' ", cleanupErr, err)
+	}
+	return err
+}
+
+func (c *magicCatalog) makeCatalogService() *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.serviceName,
+			Namespace: c.namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "grpc",
+					Port:       catalogServicePort,
+					Protocol:   "TCP",
+					TargetPort: intstr.FromInt(int(catalogServicePort)),
+				},
+			},
+			Selector: c.makeCatalogSourcePodLabels(),
+		},
+	}
+}
+
+func (c *magicCatalog) makeConfigMap() *corev1.ConfigMap {
+	isImmutable := true
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.configMapName,
+			Namespace: c.namespace,
+		},
+		Immutable: &isImmutable,
```

Comment:
Super unimportant: we likely have a test helper function (or at least we should) that returns a pointer to a bool somewhere.